File: NonTerminal.rb
# encoding: utf-8
# Rodrigo Fuentes Hernandez
# aka Blurred_9L
# Compiladores CUCEI 2013B
# NonTerminal.rb

require "./SyntaxNode.rb"
require "./NonTerminalTypes.rb"
require "./Logger.rb"
require "./RssInfo.rb"
require "./Channel.rb"

class NonTerminal < SyntaxNode
    attr_reader :children
    attr_writer :children
    
    @@ntTypes = nil
    
    private
        def visitChildren()
            for child in children
                child.visitNode()
            end
        end
    
    public
        def initialize( type, children )
            super( type )
            @children = children
        end
        
        def visitNode()
            if @type == NonTerminal.getTypes().types["channel_tag"]
                Logger.instance.writeLog( "InicioCanal\n" )
                RssInfo.instance.channels.push( Channel.new() )
            elsif @type == NonTerminal.getTypes().types["item_tag"]
                Logger.instance.writeLog( "InicioElemento\n" )
                RssInfo.instance.channels.last().foundItem()
                RssInfo.instance.inItem = true
            elsif @type == NonTerminal.getTypes().types["title_start"]
                Logger.instance.writeLog( "Titulo: " )
                if RssInfo.instance.inItem
                    RssInfo.instance.channels.last().items.last().foundTitle()
                else
                    RssInfo.instance.channels.last().foundTitle()
                end
            elsif @type == NonTerminal.getTypes().types["link_start"]
                Logger.instance.writeLog( "Enlace: " )
                if RssInfo.instance.inItem
                    RssInfo.instance.channels.last().items.last().foundLink()
                else
                    RssInfo.instance.channels.last().foundLink()
                end
            elsif @type == NonTerminal.getTypes().types["description_start"]
                Logger.instance.writeLog( "Descripcion: " )
                if RssInfo.instance.inItem
                    RssInfo.instance.channels.last().items.last().foundDescription()
                else
                    RssInfo.instance.channels.last().foundDescription()
                end
            end
            
            visitChildren()
            
            if @type == NonTerminal.getTypes().types["channel_tag"]
                Logger.instance.writeLog( "FinCanal\n" )
            elsif @type == NonTerminal.getTypes().types["item_tag"]
                Logger.instance.writeLog( "FinElemento\n" )
                RssInfo.instance.inItem = false
            elsif @type == NonTerminal.getTypes().types["title_end"] or
                  @type == NonTerminal.getTypes().types["link_end"] or    
                  @type == NonTerminal.getTypes().types["description_end"]  
                  Logger.instance.writeLog( "\n" )     
            end
        end
        
        def NonTerminal.setTypes( types )
            @@ntTypes = types
        end
        
        def NonTerminal.getTypes()
            return @@ntTypes
        end
end

File: NonTerminal.py
# encoding: utf-8
# Rodrigo Fuentes Hernandez
# aka Blurred_9L
# Compiladores CUCEI 2013B
# NonTerminal.py

import SyntaxNode
import Logger
import RssInfo
import Channel
import Item

class NonTerminal( SyntaxNode.SyntaxNode ):
    ntTypes = None
    
    def __init__( self, nodeType, children ):
        super( NonTerminal, self ).__init__( nodeType )
        self.children = children
        
    def visitNode( self, log, rssInfo ):
        if self.nodeType == getTypes().types["channel_tag"]:
            log.writeLog( "InicioCanal\n" )
            rssInfo.channels.append( Channel.Channel() )
        elif self.nodeType == getTypes().types["item_tag"]:
            log.writeLog( "InicioElemento\n" )
            rssInfo.channels[-1].foundItem()
            rssInfo.inItem = True
        elif self.nodeType == getTypes().types["title_start"]:
            log.writeLog( "Titulo:" )
            if rssInfo.inItem:
               rssInfo.channels[-1].items[-1].foundTitle( rssInfo, log )
            else:
               rssInfo.channels[-1].foundTitle( rssInfo, log )
        elif self.nodeType == getTypes().types["link_start"]:
            log.writeLog( "Enlace:" )
            if rssInfo.inItem:
               rssInfo.channels[-1].items[-1].foundLink( rssInfo, log )
            else:
               rssInfo.channels[-1].foundLink( rssInfo, log )
        elif self.nodeType == getTypes().types["description_start"]:
            log.writeLog( "Descripcion:" )
            if rssInfo.inItem:
               rssInfo.channels[-1].items[-1].foundDescription( rssInfo, log )
            else:
               rssInfo.channels[-1].foundDescription( rssInfo, log )
            
        self.visitChildren( log, rssInfo )
        
        if self.nodeType == getTypes().types["channel_tag"]:
            log.writeLog( "FinCanal\n" )
            if not rssInfo.channels[-1].isCorrect():
                log.writeError( "Error: Missing obligatory tag in channel." )
                rssInfo.errorsFound += 1
        elif self.nodeType == getTypes().types["item_tag"]:
            log.writeLog( "FinElemento\n" )
            if not rssInfo.channels[-1].items[-1].isCorrect():
                log.writeError( "Error: Missing obligatory tag in item." )
                rssInfo.errorsFound += 1
            rssInfo.inItem = False
        elif self.nodeType == getTypes().types["title_end"] or self.nodeType == getTypes().types["link_end"] or self.nodeType == getTypes().types["description_end"]:
                log.writeLog( "\n" )
                
    def visitChildren( self, log, rssInfo ):
        for child in self.children:
            child.visitNode( log, rssInfo )
        
def setTypes( types ):
    NonTerminal.ntTypes = types
    
def getTypes():
    return NonTerminal.ntTypes
